dataclients/kubernetes/crd.md

```diff
@@ -0,0 +1,634 @@
+or as simple as defining one or more redirects without having more
+than one object.  While you can use `zalando.org/skipper-routes` to do
+the redirects, we do not want to move complexity to our users.
+
+## Background Information
+
+A RouteGroup is used to store dependent group of routes in one Object,
+that can act as one thing.
+Users of this CRD are multiple controllers and end-users.
+
+### Users
+
+#### Controllers
+
+For example [stackset-controller](https://github.com/zalando-incubator/stackset-controller)
+orchestrates traffic switching. In this case a set of routes with different
+configurations to one backend should be updated atomically.
+
+Another example is an openapi-spec controller, which goal is to create
+a set of routes based on API endpoint definitions with rules, defaults
+and overrides.
+
+#### End User
+
+An end-user might want to create an Ingress with one or more
+additional paths related to the Ingress, that would do redirects.
+
+Another scenario is to refactor parts of an API into a new service.
+You want to be able to traffic switch this case, because it is common
+enough to be a considered normal case.
+
+### Goals of the RouteGroup CRD
+
+- more DRY object to use (hosts and backends separated from path rules)
+- support common use cases in a safe way (redirects, split API)
+- avoid defining unrelated route groups in the same specification object
+- better support for skipper specific features ([Predicates](https://opensource.zalando.com/skipper/reference/predicates) and [Filters](https://opensource.zalando.com/skipper/reference/filters/))
+- orchestrate traffic switching for a full set of routes (RouteGroup) without redundant configuration
+- enable complex cases: for example an enhanced openapi-spec controller write this object, instead of creating multiple ingress and [stackset-controller](https://github.com/zalando-incubator/stackset-controller) does traffic switching
+
+## Proposed solution
+
+We create a CRD with `kind: RouteGroup`, that can express the mentioned use cases.
+The spec of the RouteGroup has 3 keys that are all arrays of some kind.
+
+```yaml
+apiVersion: zalando.org/v1
+kind: RouteGroup
+spec:
+  hosts:
+  - <string>
+  backends:
+  - <backend>
+  routes:
+  - <route>
+```
+
+`<string>` is an arbitrary string
+
+`<backend>` is an object that has optional fields based on type
+
+```
+<backend>
+  name: <string>
+  type: <string> that is one of "service|shunt|loopback|lb|url"
+  url: <string> required for type=url
+  algorithm: <string> optional, only user for type=lb
+  endpoints: <stringarray> required for type=lb
+  serviceName: <string> required for type=service
+  servicePort: <number> required for type=service
+  default: <bool> optional  // TODO: maybe separate object
+  weight: <number> optional // TODO: maybe separate object
+```
+
+`<stringarray>` is an array of `<string>`
+`<number>` is an unsigned integer, example: 80
+`<bool>` is a either `true` or `false`
+
+```
+<route>
```

Comment:
One thing that I'd like to see addressed is the route specificity issue. For me as a user, it'd make sense if the routes were matched in the order they were declared in the CRD, but IIUC Skipper matches based on the number of predicates. Since a lot of these predicates will be hidden from the users, I don't think it'll be easy to reason about the ordering, which might lead to nasty surprises (e.g. a route requiring authentication silently skipped).

Reply:
This will be out of scope and not really covered. 
This would need a major change in the routing tree and I don't see too much value of this.
Also beware, that the Path and PathSubtree are reducing the routes in a lookup tree to not have to do a linear search based on predicates on all routes, but the reduced number of routes.